File: app/src/main/java/com/coinninja/coinkeeper/ui/base/AndroidActivityBuilder.java
package com.coinninja.coinkeeper.ui.base;

import com.coinninja.coinkeeper.ui.backup.BackupRecoveryWordsStartActivity;
import com.coinninja.coinkeeper.ui.phone.verification.VerifyPhoneNumberActivity;
import com.coinninja.coinkeeper.ui.settings.SettingsActivity;
import com.coinninja.coinkeeper.ui.spending.BuyBitcoinActivity;
import com.coinninja.coinkeeper.ui.spending.SpendBitcoinActivity;
import com.coinninja.coinkeeper.ui.transaction.details.TransactionDetailsActivity;
import com.coinninja.coinkeeper.ui.transaction.history.TransactionHistoryActivity;
import com.coinninja.coinkeeper.util.TwitterUtil;
import com.coinninja.coinkeeper.view.activity.AuthenticateActivity;
import com.coinninja.coinkeeper.view.activity.AuthorizedActionActivity;
import com.coinninja.coinkeeper.view.activity.BackupActivity;
import com.coinninja.coinkeeper.view.activity.BroadcastActivity;
import com.coinninja.coinkeeper.view.activity.CoinKeeperSupportActivity;
import com.coinninja.coinkeeper.view.activity.CreatePinActivity;
import com.coinninja.coinkeeper.view.activity.InviteSendActivity;
import com.coinninja.coinkeeper.view.activity.LicensesActivity;
import com.coinninja.coinkeeper.view.activity.PickContactActivity;
import com.coinninja.coinkeeper.view.activity.QrScanActivity;
import com.coinninja.coinkeeper.view.activity.RecoverWalletActivity;
import com.coinninja.coinkeeper.view.activity.RestoreWalletActivity;
import com.coinninja.coinkeeper.view.activity.SplashActivity;
import com.coinninja.coinkeeper.view.activity.StartActivity;
import com.coinninja.coinkeeper.view.activity.TrainingActivity;
import com.coinninja.coinkeeper.ui.account.verify.UserAccountVerificationActivity;
import com.coinninja.coinkeeper.view.activity.VerifyPhoneVerificationCodeActivity;
import com.coinninja.coinkeeper.view.activity.VerifyRecoverywordsActivity;
import com.coinninja.coinkeeper.view.activity.base.BalanceBarActivity;
import com.coinninja.coinkeeper.view.activity.base.MessengerActivity;
import com.coinninja.coinkeeper.view.activity.base.SecuredActivity;

import dagger.Module;
import dagger.android.ContributesAndroidInjector;

@Module()
public abstract class AndroidActivityBuilder {

    @ActivityScope
    @ContributesAndroidInjector
    abstract TwitterUtil twitterUtil();

    @ActivityScope
    @ContributesAndroidInjector
    abstract SpendBitcoinActivity spendBitcoinActivity();

    @ActivityScope
    @ContributesAndroidInjector
    abstract BuyBitcoinActivity buyBitcoinActivity();

    @ActivityScope
    @ContributesAndroidInjector
    abstract UserAccountVerificationActivity userAccountVerificationActivity();

    @ActivityScope
    @ContributesAndroidInjector
    abstract BaseActivity baseActivity();

    @ActivityScope
    @ContributesAndroidInjector
    abstract BackupActivity backupActivity();

    @ActivityScope
    @ContributesAndroidInjector
    abstract BackupRecoveryWordsStartActivity backupRecoveryWordsStartActivity();

    @ActivityScope
    @ContributesAndroidInjector
    abstract BalanceBarActivity balanceBarActivity();

    @ActivityScope
    @ContributesAndroidInjector
    abstract BroadcastActivity broadcastActivity();

    @ActivityScope
    @ContributesAndroidInjector
    abstract AuthenticateActivity authenticateActivity();

    @ActivityScope
    @ContributesAndroidInjector
    abstract AuthorizedActionActivity authorizedActionActivity();

    @ActivityScope
    @ContributesAndroidInjector
    abstract CoinKeeperSupportActivity coinKeeperSupportActivity();

    @ActivityScope
    @ContributesAndroidInjector
    abstract CreatePinActivity createPinActivity();

    @ActivityScope
    @ContributesAndroidInjector
    abstract InviteSendActivity inviteSendActivity();

    @ActivityScope
    @ContributesAndroidInjector
    abstract LicensesActivity licensesActivity();

    @ActivityScope
    @ContributesAndroidInjector
    abstract PickContactActivity pickContactActivity();

    @ActivityScope
    @ContributesAndroidInjector
    abstract MessengerActivity messagegerActivity();

    @ActivityScope
    @ContributesAndroidInjector
    abstract QrScanActivity qrScanActivity();

    @ActivityScope
    @ContributesAndroidInjector
    abstract RestoreWalletActivity restoreWalletActivity();

    @ActivityScope
    @ContributesAndroidInjector
    abstract RecoverWalletActivity recoverWalletActivity();

    @ActivityScope
    @ContributesAndroidInjector
    abstract SecuredActivity securedActivity();

    @ActivityScope
    @ContributesAndroidInjector
    abstract SettingsActivity settingsActivity();

    @ActivityScope
    @ContributesAndroidInjector
    abstract SplashActivity splashActivity();

    @ActivityScope
    @ContributesAndroidInjector
    abstract StartActivity startActivity();

    @ActivityScope
    @ContributesAndroidInjector
    abstract TransactionHistoryActivity transactionHistoryActivity();

    @ActivityScope
    @ContributesAndroidInjector
    abstract TrainingActivity trainingActivity();

    @ActivityScope
    @ContributesAndroidInjector
    abstract VerifyPhoneNumberActivity verifyPhoneNumberActivity();

    @ActivityScope
    @ContributesAndroidInjector
    abstract VerifyPhoneVerificationCodeActivity verifyPhoneVerificationCodeActivity();

    @ActivityScope
    @ContributesAndroidInjector
    abstract VerifyRecoverywordsActivity verifyRecoverywordsActivity();

    @ActivityScope
    @ContributesAndroidInjector
    abstract TransactionDetailsActivity transactionDetailsActivity();

}

File: app/src/main/java/com/coinninja/coinkeeper/di/builder/AndroidActivityBuilder.kt
package com.coinninja.coinkeeper.di.builder

import com.coinninja.coinkeeper.di.interfaces.ActivityScope
import com.coinninja.coinkeeper.ui.account.verify.UserAccountVerificationActivity
import com.coinninja.coinkeeper.ui.backup.BackupRecoveryWordsStartActivity
import com.coinninja.coinkeeper.ui.base.BaseActivity
import com.coinninja.coinkeeper.ui.base.BaseActivityModule
import com.coinninja.coinkeeper.ui.home.HomeActivity
import com.coinninja.coinkeeper.ui.home.HomeModule
import com.coinninja.coinkeeper.ui.lightning.broadcast.BroadcastLightningPaymentActivity
import com.coinninja.coinkeeper.ui.lightning.deposit.LightningDepositActivity
import com.coinninja.coinkeeper.ui.lightning.deposit.LightningDepositActivityModule
import com.coinninja.coinkeeper.ui.lightning.withdrawal.LightningWithdrawalActivity
import com.coinninja.coinkeeper.ui.lightning.withdrawal.LightningWithdrawalActivityModule
import com.coinninja.coinkeeper.ui.lightning.withdrawal.LightningWithdrawalBroadcastActivity
import com.coinninja.coinkeeper.ui.market.MarketChartModule
import com.coinninja.coinkeeper.ui.market.MarketScreenActivity
import com.coinninja.coinkeeper.ui.news.MarketNewsModule
import com.coinninja.coinkeeper.ui.payment.confirm.ConfirmPaymentActivity
import com.coinninja.coinkeeper.ui.payment.create.CreatePaymentActivity
import com.coinninja.coinkeeper.ui.payment.create.CreatePaymentModule
import com.coinninja.coinkeeper.ui.payment.invite.InviteContactActivity
import com.coinninja.coinkeeper.ui.payment.request.LndInvoiceRequestActivity
import com.coinninja.coinkeeper.ui.payment.request.PayRequestActivity
import com.coinninja.coinkeeper.ui.payment.request.PayRequestScreenModule
import com.coinninja.coinkeeper.ui.phone.verification.VerificationActivity
import com.coinninja.coinkeeper.ui.segwit.*
import com.coinninja.coinkeeper.ui.settings.AdjustableFeesActivity
import com.coinninja.coinkeeper.ui.settings.SettingsActivity
import com.coinninja.coinkeeper.ui.spending.BuyBitcoinActivity
import com.coinninja.coinkeeper.ui.spending.SpendBitcoinActivity
import com.coinninja.coinkeeper.ui.transaction.details.TransactionDetailsActivity
import com.coinninja.coinkeeper.view.activity.*
import dagger.Module
import dagger.android.ContributesAndroidInjector

@Module
abstract class AndroidActivityBuilder {

    @ActivityScope
    @ContributesAndroidInjector(modules = [HomeModule::class, BaseActivityModule::class])
    internal abstract fun homeActivity(): HomeActivity

    @ActivityScope
    @ContributesAndroidInjector(modules = [BaseActivityModule::class, MarketChartModule::class, MarketNewsModule::class])
    internal abstract fun marketScreenActivity(): MarketScreenActivity

    @ActivityScope
    @ContributesAndroidInjector(modules = [BaseActivityModule::class])
    internal abstract fun adjustableFeesActivity(): AdjustableFeesActivity

    @ActivityScope
    @ContributesAndroidInjector(modules = [BaseActivityModule::class])
    internal abstract fun signupSelectionActivity(): SignUpSelectionActivity

    @ActivityScope
    @ContributesAndroidInjector(modules = [BaseActivityModule::class])
    internal abstract fun spendBitcoinActivity(): SpendBitcoinActivity

    @ActivityScope
    @ContributesAndroidInjector(modules = [BaseActivityModule::class])
    internal abstract fun buyBitcoinActivity(): BuyBitcoinActivity

    @ActivityScope
    @ContributesAndroidInjector(modules = [BaseActivityModule::class])
    internal abstract fun userAccountVerificationActivity(): UserAccountVerificationActivity

    @ActivityScope
    @ContributesAndroidInjector(modules = [BaseActivityModule::class])
    internal abstract fun baseActivity(): BaseActivity

    @ActivityScope
    @ContributesAndroidInjector(modules = [BaseActivityModule::class])
    internal abstract fun backupActivity(): BackupActivity

    @ActivityScope
    @ContributesAndroidInjector(modules = [BaseActivityModule::class])
    internal abstract fun backupRecoveryWordsStartActivity(): BackupRecoveryWordsStartActivity

    @ActivityScope
    @ContributesAndroidInjector(modules = [BaseActivityModule::class])
    internal abstract fun broadcastActivity(): BroadcastActivity

    @ActivityScope
    @ContributesAndroidInjector(modules = [BaseActivityModule::class])
    internal abstract fun authenticateActivity(): AuthenticateActivity

    @ActivityScope
    @ContributesAndroidInjector(modules = [BaseActivityModule::class])
    internal abstract fun authorizedActionActivity(): AuthorizedActionActivity

    @ActivityScope
    @ContributesAndroidInjector(modules = [BaseActivityModule::class])
    internal abstract fun coinKeeperSupportActivity(): CoinKeeperSupportActivity

    @ActivityScope
    @ContributesAndroidInjector(modules = [BaseActivityModule::class])
    internal abstract fun createPinActivity(): CreatePinActivity

    @ActivityScope
    @ContributesAndroidInjector(modules = [BaseActivityModule::class])
    internal abstract fun inviteSendActivity(): InviteSendActivity

    @ActivityScope
    @ContributesAndroidInjector(modules = [BaseActivityModule::class])
    internal abstract fun licensesActivity(): LicensesActivity

    @ActivityScope
    @ContributesAndroidInjector(modules = [BaseActivityModule::class])
    internal abstract fun pickContactActivity(): PickUserActivity

    @ActivityScope
    @ContributesAndroidInjector(modules = [BaseActivityModule::class])
    internal abstract fun qrScanActivity(): QrScanActivity

    @ActivityScope
    @ContributesAndroidInjector(modules = [BaseActivityModule::class])
    internal abstract fun restoreWalletActivity(): RestoreWalletActivity

    @ActivityScope
    @ContributesAndroidInjector(modules = [BaseActivityModule::class])
    internal abstract fun recoverWalletActivity(): RecoverWalletActivity

    @ActivityScope
    @ContributesAndroidInjector(modules = [BaseActivityModule::class])
    internal abstract fun settingsActivity(): SettingsActivity

    @ActivityScope
    @ContributesAndroidInjector(modules = [BaseActivityModule::class])
    internal abstract fun splashActivity(): SplashActivity

    @ActivityScope
    @ContributesAndroidInjector(modules = [BaseActivityModule::class])
    internal abstract fun startActivity(): StartActivity

    @ActivityScope
    @ContributesAndroidInjector(modules = [BaseActivityModule::class])
    internal abstract fun trainingActivity(): TrainingActivity

    @ActivityScope
    @ContributesAndroidInjector(modules = [BaseActivityModule::class])
    internal abstract fun verifyPhoneNumberActivity(): VerificationActivity

    @ActivityScope
    @ContributesAndroidInjector(modules = [BaseActivityModule::class])
    internal abstract fun verifyPhoneVerificationCodeActivity(): VerifyPhoneVerificationCodeActivity

    @ActivityScope
    @ContributesAndroidInjector(modules = [BaseActivityModule::class])
    internal abstract fun verifyRecoverywordsActivity(): VerifyRecoverywordsActivity

    @ActivityScope
    @ContributesAndroidInjector(modules = [BaseActivityModule::class])
    internal abstract fun transactionDetailsActivity(): TransactionDetailsActivity

    @ActivityScope
    @ContributesAndroidInjector(modules = [BaseActivityModule::class, LightningDepositActivityModule::class])
    internal abstract fun lightningDepositActivity(): LightningDepositActivity

    @ActivityScope
    @ContributesAndroidInjector(modules = [BaseActivityModule::class, LightningWithdrawalActivityModule::class])
    internal abstract fun lightningWithdrawalActivity(): LightningWithdrawalActivity

    @ActivityScope
    @ContributesAndroidInjector(modules = [BaseActivityModule::class, LightningWithdrawalActivityModule::class])
    internal abstract fun lightningWithdrawalCompletedActivity(): LightningWithdrawalBroadcastActivity

    @ActivityScope
    @ContributesAndroidInjector(modules = [BaseActivityModule::class, UpgradeToSegwitActivityModule::class])
    internal abstract fun upgradeToSegwitActivity(): UpgradeToSegwitActivity

    @ActivityScope
    @ContributesAndroidInjector(modules = [BaseActivityModule::class, PerformSegwitUpgradeActivityModule::class])
    internal abstract fun performSegwitUpgradeActivity(): PerformSegwitUpgradeActivity

    @ActivityScope
    @ContributesAndroidInjector(modules = [BaseActivityModule::class])
    internal abstract fun upgradeToSegwitCompleteActivity(): UpgradeToSegwitCompleteActivity

    @ActivityScope
    @ContributesAndroidInjector(modules = [BaseActivityModule::class, PayRequestScreenModule::class])
    internal abstract fun payRequestActivity(): PayRequestActivity

    @ActivityScope
    @ContributesAndroidInjector(modules = [BaseActivityModule::class, PayRequestScreenModule::class])
    internal abstract fun lndInvoiceRequestActivity(): LndInvoiceRequestActivity

    @ActivityScope
    @ContributesAndroidInjector(modules = [BaseActivityModule::class, CreatePaymentModule::class])
    internal abstract fun createPaymentActivity(): CreatePaymentActivity

    @ActivityScope
    @ContributesAndroidInjector(modules = [BaseActivityModule::class, CreatePaymentModule::class])
    internal abstract fun confirmPaymentActivity(): ConfirmPaymentActivity

    @ActivityScope
    @ContributesAndroidInjector(modules = [BaseActivityModule::class, CreatePaymentModule::class])
    internal abstract fun broadcastLightningPaymentActivity(): BroadcastLightningPaymentActivity

    @ActivityScope
    @ContributesAndroidInjector(modules = [BaseActivityModule::class, CreatePaymentModule::class])
    internal abstract fun inviteContactActivity(): InviteContactActivity
}
